# My First Hamoye Quiz

Tag-along codes for questions on "FoodBalanceSheets_E_Africa_NOFLAG.csv" dataset.

In [112]:
import pandas as pd

In [10]:
df = pd.read_csv("FoodBalanceSheets_E_Africa_NOFLAG.csv", encoding = "latin-1")

### 1. Select columns ‘Y2017’ and ‘Area’, Perform a groupby operation on ‘Area’. Which of these Areas had the 7th lowest sum in 2017?


In [60]:
area_2017=df[["Y2017", "Area"]].groupby("Area").sum().sort_values(by= "Y2017", ascending=True).reset_index()
area_2017

Area       Y2017
0                Sudan (former)        0.00
1                  Ethiopia PDR        0.00
2                       Comoros       59.84
3                    Seychelles      442.34
4         Sao Tome and Principe    12662.63
5                    Cabo Verde    14650.74
6                 Guinea-Bissau    19102.77
7                       Lesotho    21267.96
8                      Botswana    22101.30
9                      Djibouti    22729.91
10                       Gambia    23154.18
11                        Gabon    27979.64
12                      Liberia    29342.20
13                      Namibia    29874.89
14     Central African Republic    29937.00
15                        Congo    41181.68
16                         Togo    49841.88
17                    Mauritius    51114.83
18                     Eswatini    54343.33
19                 Sierra Leone    55311.33
20                         Chad    71594.68
21                       Rwanda    73663.69
22                     Zimbabwe    75919.34
23                      Senegal    95681.15
24                       Guinea    98138.87
25                 Burkina Faso   101855.07
26                       Zambia   103223.77
27                      Tunisia   124167.20
28                        Benin   124771.22
29                        Niger   126707.58
30                   Madagascar   131197.73
31                         Mali   149928.33
32                   Mauritania   156665.46
33                   Mozambique   161407.98
34                       Malawi   181098.71
35                       Uganda   213950.38
36                Côte d'Ivoire   224599.01
37                       Angola   229159.57
38                     Cameroon   232030.43
39                        Sudan   239931.92
40                        Kenya   264660.66
41  United Republic of Tanzania   322616.85
42                      Algeria   325644.27
43                        Ghana   337599.06
44                      Morocco   388495.36
45                     Ethiopia   448683.76
46                 South Africa   517590.54
47                        Egypt   866379.92
48                      Nigeria  1483268.23

Areas were sorted from lowest to highest by sum per 2017. In the 7th row(index is 6) there is Guinea-Bissau. So, **Guinea-Bissau** had the 7th lowest sum in 2017.

### 2. Select columns ‘Y2017’ and ‘Area’, Perform a groupby operation on ‘Area’.  Which of these Areas had the highest sum in 2017?

In [110]:
area_2017=df[["Y2017", "Area"]].groupby("Area").sum().reset_index()
area_2017

Area       Y2017
0                       Algeria   325644.27
1                        Angola   229159.57
2                         Benin   124771.22
3                      Botswana    22101.30
4                  Burkina Faso   101855.07
5                    Cabo Verde    14650.74
6                      Cameroon   232030.43
7      Central African Republic    29937.00
8                          Chad    71594.68
9                       Comoros       59.84
10                        Congo    41181.68
11                Côte d'Ivoire   224599.01
12                     Djibouti    22729.91
13                        Egypt   866379.92
14                     Eswatini    54343.33
15                     Ethiopia   448683.76
16                 Ethiopia PDR        0.00
17                        Gabon    27979.64
18                       Gambia    23154.18
19                        Ghana   337599.06
20                       Guinea    98138.87
21                Guinea-Bissau    19102.77
22                        Kenya   264660.66
23                      Lesotho    21267.96
24                      Liberia    29342.20
25                   Madagascar   131197.73
26                       Malawi   181098.71
27                         Mali   149928.33
28                   Mauritania   156665.46
29                    Mauritius    51114.83
30                      Morocco   388495.36
31                   Mozambique   161407.98
32                      Namibia    29874.89
33                        Niger   126707.58
34                      Nigeria  1483268.23
35                       Rwanda    73663.69
36        Sao Tome and Principe    12662.63
37                      Senegal    95681.15
38                   Seychelles      442.34
39                 Sierra Leone    55311.33
40                 South Africa   517590.54
41                        Sudan   239931.92
42               Sudan (former)        0.00
43                         Togo    49841.88
44                      Tunisia   124167.20
45                       Uganda   213950.38
46  United Republic of Tanzania   322616.85
47                       Zambia   103223.77
48                     Zimbabwe    75919.34

In [111]:
max_area_2017 = area_2017[area_2017["Y2017"] == area_2017["Y2017"].max()]
max_area_2017

Area       Y2017
34  Nigeria  1483268.23

So, **Nigeria** had the highest sum in 2017.

### 8. What is the total number of unique countries in the dataset?


In [76]:
unique_countries = df['Area'].nunique()
unique_countries

49

### 9. What is the total number and percentage of missing data in 2014 to 3 decimal places?


In [78]:
# Find total number of missind data in 2014
missing_2014 = df['Y2014'].isnull().sum()
missing_2014

1589

In [81]:
# Total number of rows in the DataFrame
total_rows = len(df)
# Find percentage of missing values in 2014
missing_2014_perc = (missing_2014 / total_rows) * 100
missing_2014_perc

2.6073544131401474

So, the total number and percentage of missing data in 2014 are **1589** and **2.607** respectively.

### 11. Perform a groupby operation on ‘Element’.  What year has the highest sum of Stock Variation?


In [85]:
elements = df.groupby("Element").sum()
elements

Area Code  \
Element                                             
Domestic supply quantity                   708993   
Export Quantity                            599910   
Fat supply quantity (g/capita/day)         675050   
Feed                                       176272   
Food                                       663295   
Food supply (kcal/capita/day)              674057   
Food supply quantity (kg/capita/yr)        658446   
Import Quantity                            688174   
Losses                                     274353   
Other uses (non-food)                      235554   
Processing                                 271940   
Production                                 526751   
Protein supply quantity (g/capita/day)     675050   
Residuals                                  623271   
Seed                                       103537   
Stock Variation                            571566   
Total Population - Both sexes                6020   
Tourist consumption                         50308   

                                                                                     Area  \
Element                                                                                     
Domestic supply quantity                AlgeriaAlgeriaAlgeriaAlgeriaAlgeriaAlgeriaAlge...   
Export Quantity                         AlgeriaAlgeriaAlgeriaAlgeriaAlgeriaAlgeriaAlge...   
Fat supply quantity (g/capita/day)      AlgeriaAlgeriaAlgeriaAlgeriaAlgeriaAlgeriaAlge...   
Feed                                    AlgeriaAlgeriaAlgeriaAlgeriaAlgeriaAlgeriaAlge...   
Food                                    AlgeriaAlgeriaAlgeriaAlgeriaAlgeriaAlgeriaAlge...   
Food supply (kcal/capita/day)           AlgeriaAlgeriaAlgeriaAlgeriaAlgeriaAlgeriaAlge...   
Food supply quantity (kg/capita/yr)     AlgeriaAlgeriaAlgeriaAlgeriaAlgeriaAlgeriaAlge...   
Import Quantity                         AlgeriaAlgeriaAlgeriaAlgeriaAlgeriaAlgeriaAlge...   
Losses                                  AlgeriaAlgeriaAlgeriaAlgeriaAlgeriaAlgeriaAlge...   
Other uses (non-food)                   AlgeriaAlgeriaAlgeriaAlgeriaAlgeriaAlgeriaAlge...   
Processing                              AlgeriaAlgeriaAlgeriaAlgeriaAlgeriaAlgeriaAlge...   
Production                              AlgeriaAlgeriaAlgeriaAlgeriaAlgeriaAlgeriaAlge...   
Protein supply quantity (g/capita/day)  AlgeriaAlgeriaAlgeriaAlgeriaAlgeriaAlgeriaAlge...   
Residuals                               AlgeriaAlgeriaAlgeriaAlgeriaAlgeriaAlgeriaAlge...   
Seed                                    AlgeriaAlgeriaAlgeriaAlgeriaAlgeriaAlgeriaAlge...   
Stock Variation                         AlgeriaAlgeriaAlgeriaAlgeriaAlgeriaAlgeriaAlge...   
Total Population - Both sexes           AlgeriaAngolaBeninBotswanaBurkina FasoCabo Ver...   
Tourist consumption                     AngolaAngolaBotswanaBotswanaBotswanaBotswanaBo...   

                                        Item Code  \
Element                                             
Domestic supply quantity                 14197445   
Export Quantity                          11840553   
Fat supply quantity (g/capita/day)       13535000   
Feed                                      3538507   
Food                                     13285035   
Food supply (kcal/capita/day)            13511060   
Food supply quantity (kg/capita/yr)      13185401   
Import Quantity                          13795966   
Losses                                    5424803   
Other uses (non-food)                     4729749   
Processing                                5350416   
Production                               10450053   
Protein supply quantity (g/capita/day)   13535000   
Residuals                                12421089   
Seed                                      2035933   
Stock Variation                          11329527   
Total Population - Both sexes              112545   
Tourist consumption                       1486528   

                                                         

In [86]:
# Filter rows where 'Element' is 'Stock Variation'
stock_variation = elements[elements.index == 'Stock Variation']
stock_variation

Area Code                                               Area  \
Element                                                                         
Stock Variation     571566  AlgeriaAlgeriaAlgeriaAlgeriaAlgeriaAlgeriaAlge...   

                 Item Code                                               Item  \
Element                                                                         
Stock Variation   11329527  Cereals - Excluding BeerWheat and productsRice...   

                 Element Code  \
Element                         
Stock Variation      21464704   

                                                              Unit     Y2014  \
Element                                                                        
Stock Variation  1000 tonnes1000 tonnes1000 tonnes1000 tonnes10...  58749.83   

                    Y2015     Y2016     Y2017     Y2018  
Element                                                  
Stock Variation  34910.99  33140.12  54316.91  20577.91

We see from the table, that the highest sum of Stock Variation is in column Y2014 - value 58749.83. So, the answer is **2014**.

### 14. Perform a groupby operation on ‘Element’.  What is the total number of the sum of Processing in 2017?

In [88]:
elements_2017 = df.groupby('Element')['Y2017'].sum().reset_index()
elements_2017


Element       Y2017
0                 Domestic supply quantity  2088198.10
1                          Export Quantity   182338.80
2       Fat supply quantity (g/capita/day)    10253.84
3                                     Feed   223705.68
4                                     Food  1258888.28
5            Food supply (kcal/capita/day)   454681.00
6      Food supply quantity (kg/capita/yr)    48690.04
7                          Import Quantity   294559.09
8                                   Losses   160614.00
9                    Other uses (non-food)    91645.97
10                              Processing   292836.00
11                              Production  2030056.89
12  Protein supply quantity (g/capita/day)    11842.45
13                               Residuals    35500.00
14                                    Seed    24870.14
15                         Stock Variation    54316.91
16           Total Population - Both sexes  1112641.00
17                     Tourist consumption       91.00

We see Processing in row 10, so, the total number of the sum of Processing in 2017 is **292836**.

### 15. Which year had the least correlation with ‘Element Code’?

In [89]:
df.corr(numeric_only = True)

Area Code  Item Code  Element Code     Y2014     Y2015  \
Area Code      1.000000  -0.005159     -0.000209  0.006164  0.005472   
Item Code     -0.005159   1.000000     -0.024683  0.021722  0.020857   
Element Code  -0.000209  -0.024683      1.000000  0.024457  0.023889   
Y2014          0.006164   0.021722      0.024457  1.000000  0.994647   
Y2015          0.005472   0.020857      0.023889  0.994647  1.000000   
Y2016          0.005247   0.020109      0.023444  0.996081  0.995739   
Y2017          0.005006   0.021494      0.024254  0.995230  0.988048   
Y2018          0.005665   0.021314      0.024279  0.994872  0.988208   

                 Y2016     Y2017     Y2018  
Area Code     0.005247  0.005006  0.005665  
Item Code     0.020109  0.021494  0.021314  
Element Code  0.023444  0.024254  0.024279  
Y2014         0.996081  0.995230  0.994872  
Y2015         0.995739  0.988048  0.988208  
Y2016         1.000000  0.992785  0.992757  
Y2017         0.992785  1.000000  0.998103  
Y2018         0.992757  0.998103  1.000000

The least correlation with ‘Element Code’ is in **Y2016** (value 0.023444).

### 17. What is the mean and standard deviation across the whole dataset for the year 2017 to 2 decimal places?

In [90]:
df.describe()

Area Code     Item Code  Element Code          Y2014          Y2015  \
count  60943.000000  60943.000000  60943.000000   59354.000000   59395.000000   
mean     134.265576   2687.176706   3814.856456     134.196282     135.235966   
std       72.605709    146.055739   2212.007033    1567.663696    1603.403984   
min        4.000000   2501.000000    511.000000   -1796.000000   -3161.000000   
25%       74.000000   2562.000000    684.000000       0.000000       0.000000   
50%      136.000000   2630.000000   5142.000000       0.090000       0.080000   
75%      195.000000   2775.000000   5511.000000       8.340000       8.460000   
max      276.000000   2961.000000   5911.000000  176405.000000  181137.000000   

               Y2016          Y2017          Y2018  
count   59408.000000   59437.000000   59507.000000  
mean      136.555222     140.917765     143.758381  
std      1640.007194    1671.862359    1710.782658  
min     -3225.000000   -1582.000000   -3396.000000  
25%         0.000000       0.000000       0.000000  
50%         0.080000       0.100000       0.070000  
75%         8.430000       9.000000       9.000000  
max    185960.000000  190873.000000  195875.000000

.describe method gives us info about the mean and standard deviation across the whole dataset. In year 2017 they are **140.92**	and **1671.86**.

### 18. What is the total Protein supply quantity in Madagascar in 2015?


In [94]:
df[(df["Area"] == "Madagascar") & (df["Element"] == "Protein supply quantity (g/capita/day)")].groupby("Element").sum()

Area Code  \
Element                                             
Protein supply quantity (g/capita/day)      14319   

                                                                                     Area  \
Element                                                                                     
Protein supply quantity (g/capita/day)  MadagascarMadagascarMadagascarMadagascarMadaga...   

                                        Item Code  \
Element                                             
Protein supply quantity (g/capita/day)     299405   

                                                                                     Item  \
Element                                                                                     
Protein supply quantity (g/capita/day)  Grand TotalVegetal ProductsAnimal ProductsCere...   

                                        Element Code  \
Element                                                
Protein supply quantity (g/capita/day)         74814   

                                                                                     Unit  \
Element                                                                                     
Protein supply quantity (g/capita/day)  g/capita/dayg/capita/dayg/capita/dayg/capita/d...   

                                         Y2014   Y2015  Y2016   Y2017   Y2018  
Element                                                                        
Protein supply quantity (g/capita/day)  177.79  173.05  173.4  175.75  176.64

In 2015 the total Protein supply quantity in Madagascar was **173.05**.

### 20. Answer the following questions based on the African food production dataset provided by the FAO website already provided. What is the total sum of Wine produced in 2015 and 2018 respectively?

Hint: Perform a groupby sum aggregation on ‘Item’

In [97]:
df[["Item", "Y2015", "Y2018"]].groupby("Item").sum()

Y2015      Y2018
Item                                     
Alcohol, Non-Food      2180.00    2293.00
Alcoholic Beverages   98783.72   97847.27
Animal Products       11811.73   11578.61
Animal fats          200675.72  269648.27
Apples and products   10559.15    9640.51
...                        ...        ...
Vegetables, Other    158104.08  163987.21
Vegetal Products     107064.17  107775.39
Wheat and products   234710.51  242645.19
Wine                   4251.81    4039.32
Yams                 203151.78  221272.09

[119 rows x 2 columns]

The total sum of Wine produced in 2015 was **4251.81**,  and in 2018 it was **4039.32**.